### DISCUSSION TITLE
Allow access to the props in next/dynamic loading render function

### DISCUSSION THREAD

**Discussion by Janpot** [vercel/next.js#26318] [Ideas]
### Describe the feature you'd like to request

Trying to render a code highlighter using `next/dynamic`, would like to offer a fallback of non-highlighted code during loading of the component. For example:

```tsx
const MyCodeHighlighter = dynamic(
  () => import('./MyCodeHighlighter'),
  { loading: () => 'loading...' }
)

// ...

<MyCodeHighlighter code='console.log("Highlight me when you can")' />
```

### Describe the solution you'd like

Pass the props to the `loading` render function so that I can show a true fallback
```tsx
const MyCodeHighlighter = dynamic(
  () => import('./MyCodeHighlighter'),
  { loading: ({ props }) => <pre>{props.code}</pre> }
)
```

### Describe alternatives you've considered

There are no alternatives that I'm aware of.

**Comment by timneutkens** [comment]
Ideally I'd like to avoid making changes to `next/dynamic` as it'll be completely revamped with React 18.

**Reply by Janpot** [reply]
Yeah, I figured server components would supersede this. For the record and my own understanding though, there currently isn't a pattern with `next/dynamic` that allows for this behavior, right?

**Reply by timneutkens** [reply]
Potentially it works through context

**Reply by osdiab** [reply]
Are you referring to using a context provider around where the dynamic component is used, and then consuming that context in the fallback? kinda jank but i guess that works lol. Would it be bad to just instantiate the dynamic component inside the component that would use it (where the props are available)? also jank but guess if next React will obviate the need for this, we can put up with it lol.

**Reply by Janpot** [reply]
@osdiab fwiw, this seems to work for me:

```tsx
// ./lib/dynamic.jsx
import dynamic from "next/dynamic";
import * as React from "react";

export default function dynamicWithFallbackProps(getter, options = {}) {
  const PropsContext = React.createContext({});
  const patchedOptions = options.loading
    ? {
        ...options,
        loading: function Loading(...props) {
          const innerProps = React.useContext(PropsContext);
          return React.createElement(options.loading, {
            ...props,
            props: innerProps
          });
        }
      }
    : options;
  const Inner = dynamic(getter, patchedOptions);
  return function Component(props) {
    return (
      <PropsContext.Provider value={props}>
        <Inner {...props} />
      </PropsContext.Provider>
    );
  };
}

// ./pages/index.jsx
import dynamic from "../lib/dynamic";
import * as React from "react";

const HighlightedCode = dynamic(() => import("../components/HighlightedCode"), {
  loading: ({ props }) => <div>fallback: {props.code}</div>
});
```

**Reply by osdiab** [reply]
ah cool makes more sense now that it's concrete, thanks for the useful workaround!
